crates/store/re_log_encoding/src/decoder/stream.rs

```diff
@@ -221,14 +351,85 @@ impl StreamDecoder {
     }
 }
 
-fn propagate_version(message: &mut LogMsg, version: Option<CrateVersion>) {
-    if let re_log_types::LogMsg::SetStoreInfo(msg) = message {
-        // Propagate the protocol version from the header into the `StoreInfo` so that all
-        // parts of the app can easily access it.
-        msg.info.store_version = version;
+// ---
+
+/// Iteratively decodes the contents of an arbitrary buffered reader.
+pub struct StreamDecoderIterator<T: FileEncoded, R: std::io::BufRead> {
+    decoder: StreamDecoder<T>,
+    reader: R,
+
+    /// If true, the decoder will always wait for an end-of-stream marker before calling it a day,
+    /// even if the underlying reader has nothing to yield.
+    wait_for_eos: bool,
+
+    /// See [`StreamDecoder::decode_eager`] for more information.
+    first_msg: Option<T>,
+}
+
+impl<T: FileEncoded, R: std::io::BufRead> StreamDecoderIterator<T, R> {
+    pub fn num_bytes_processed(&self) -> u64 {
+        self.decoder.byte_chunks.num_read() as _
     }
 }
 
+impl<T: FileEncoded, R: std::io::BufRead> std::iter::Iterator for StreamDecoderIterator<T, R> {
+    type Item = Result<T, DecodeError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(first_msg) = self.first_msg.take() {
+            // The iterator was eagerly initialized so make sure to return the first message if there's any.
+            return Some(Ok(first_msg));
+        }
+
+        loop {
+            match self.decoder.try_read() {
+                Ok(Some(msg)) => return Some(Ok(msg)),
+                Ok(None) => {}
+                Err(err) => return Some(Err(err)),
+            }
+
+            match self.reader.fill_buf() {
+                // EOF
+                Ok([]) => {
+                    // There's nothing more to read…
+                    match self.decoder.try_read() {
+                        // …but we still have enough buffered that we can still manage to decode
+                        // more messages, so go on for now.
+                        Ok(Some(msg)) => return Some(Ok(msg)),
+
+                        // …and we're not expecting concatenated streams, so just leave.
```

Comment:
Conflating concatenated streams with EOS markers is the source of major confusion, let's not repeat that here.